File: concur.clj
(let [chan (java.util.concurrent.LinkedBlockingQueue.)
      result (future (let [a (.take chan)
                           b (.take chan)
                           c (.take chan)]
                       (str a " + " b " + " c " = " (+ a b c))))]
  (future (.put chan (* 2 10)))
  (future (.put chan (* 2 20)))
  (future (.put chan (+ 30 40)))
  (println @result))

File: concur.kt
import java.util.concurrent.LinkedBlockingQueue
import kotlin.concurrent.thread

fun main(args : Array<String>) {
  val numq = LinkedBlockingQueue<Int>()
  val strq = LinkedBlockingQueue<String>()
  thread() {
    val a = numq.take()
    val b = numq.take()
    val c = numq.take()
    val d = a?.plus(b!!)?.plus(c!!)
    strq.put("$a + $b + $c = $d")
  }
  thread() { numq.put(2 * 10) }
  thread() { numq.put(2 * 20) }
  thread() { numq.put(30 + 40) }
  println(strq.take())
}
